Note that path state cannot be removed without reception of PATH_ABANDON (#475)
quicwg/multipath: draft-ietf-quic-multipath.md

{{spurious-stateless-reset}} and to be able to acknowledge the
last packets received from the peer as specified in {{ack-after-abandon}}.

If a peer sends a PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it might not be able to remove path state.
It is left to the implementation to handle this unexpected

> behavior as it does not impact interoperability.

Author: ```suggestion
behavior as it does not impact interoperability. However, an endpoint SHOULD NOT
remove path state earlier than 3 PTOs after sending the PATh_ABANDON frame.
```

{{spurious-stateless-reset}} and to be able to acknowledge the
last packets received from the peer as specified in {{ack-after-abandon}}.

If a peer sends a PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it might not be able to remove path state.
It is left to the implementation to handle this unexpected

> behavior as it does not impact interoperability. However, an endpoint SHOULD NOT
> remove path state earlier than 3 PTOs after sending the PATh_ABANDON frame.

Author: ```suggestion
behavior.  Implementation MAY shed some of the path state and reduce
resource consumption if the PATH_ABANDON is
not arriving in a reasonable time, but they SHOULD
take steps and avoid sending spurious stateless reset (see {{spurious-stateless-reset}}).
```

> remove path state earlier than 3 PTOs after sending the PATh_ABANDON frame.

Author: ```suggestion
remove issued CIDs earlier than 3 PTOs after sending the PATh_ABANDON frame.
```

Reviewer: There is no way to "(retire) the CIDs that you provided to the peer." Only the peer can do that, either by sending individual RETIRE_CONNECTION_ID frames or by sending a PATH_ABANDON frame.

Of course, you can decide to unilaterally "forget" these CIDs. The practical effect is that if the peer uses them to send packets, the forgotten CIDs will trigger a stateless reset, thus closing the connection. If you want to do that, you can just as well close the connection with some appropriate error code. A frank closure of the connection seems better than a passive-aggressive booby trap...

Maybe we should define an error code for that scenario. "I waited for the path abandon for more that 3 PTOs, and did not receive the expected path abandon, so bye. Try to fix your bugs next time."

Reviewer: Actually, you could send a "NEW CONNECTION ID" with a "retire prior", but that's bad because you have abandoned the path already? Plus, the new connection ID will have the path ID of the abandoned path, so that does not let you abandon the path ID.

Author: Actually RFC9000 says:

> "An endpoint might need to stop accepting previously issued connection IDs in certain circumstances. Such an endpoint can cause its peer to retire connection IDs by sending a NEW_CONNECTION_ID frame with an increased Retire Prior To field. The endpoint SHOULD continue to accept the previously issued connection IDs until they are retired by the peer. If the endpoint can no longer process the indicated connection IDs, it MAY close the connection."

So effectively even a "NEW CONNECTION ID" with a "retire prior" will not retire them. And given the text above I guess closing the connection is the right/only option. Therefore see proposed new text aligning with language in RFC9000.

I'm still proposing this PR because I still think we should spell out normatively that you should wait at least 3 PTO before doing anything. Given this is not enforceable I would stick with a SHOULD but effectively this is meant to be a MUST.

> behavior as it does not impact interoperability. However, an endpoint SHOULD NOT
> remove path state earlier than 3 PTOs after sending the PATh_ABANDON frame.

Author: ```suggestion
behavior as it does not impact interoperability. If the endpoint is no longer 
willing to process the issued connection IDs for the abandoned path,
it MAY close the connection, but SHOULD wait at least 3 PTOs after
sending the PATh_ABANDON frame.
```

Reviewer: Yes, that works. It avoids saying creative things, and it sends the right message to developers: reply by ABANDON to ABANDON, or risk losing the connection. Note the typo "PATh_ABANDON".

Reviewer 2: I believe it works.

Reviewer 3: Yes, this works for me.

> 

Reviewer: What state are we speaking about? I think that we are saying either too much or too little. There many kinds of "states":

1) Lists of packets that were sent on the path and not yet acked by the peer,
2) Packets that were received from the peer and processed but not yet acked,
3) the total number of paths considered active, which some implementation use to decide when to increase the Max Path ID.

Suppose that an implementation has abandoned a path, had not received a Path Abandon from the peer, and is losing patience. It may do a number of things:

1) It may decide to drop any incoming packet on that path. That has no interoperability impact, as long as those dropped packets are never acknowledged. Indeed, that's exactly what would happen if the path had become bad and packets were dropped in the network. But it has some performance impact, since some packets will be in transit when the Path Abandon is sent. It is recommended to not do that earlier than 3 PTOs after sending the PATH_ABANDON frame.

2) It may decide to consider all packets send on that path as lost, and resend the frames through another path. That's fine. It might cause some duplicate transmission, but that also happens for other reasons (e.g., bad timers) and has no interoperability impact. The rules there are murky, because this is a tradeoff -- resend sooner minimizes the end-to-end latency caused by losses, but resend sooner also increases the risk of duplicate transmission, which reduces efficiency. The 3PTO rule is probably one end of the tradeoff: good mitigation of duplicate transmission, poor management of latency. A 1PTO rule would be a different tradeoff, more focused on managing latency.

3) It should send the "missing ACKs" on another path before clearing the missing ACK state. Again, that's an efficiency tradeoff. If the node does not send acknowledgements, it risks causing spurious retransmission. 

4) It may well decide to never increase the MAX_PATH_ID until the PATH_ABANDON has been received. That's a reasonable way for an implementation to conserve resource.

Reviewer: I would just say:

If a peer sends a PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it might not be able to remove path state.
It is left to the implementation to handle this unexpected
behavior.

Author: > What state are we speaking about? I think that we are saying either too much or too little. There many kinds of "states":
> 
> 1. Lists of packets that were sent on the path and not yet acked by the peer,
> 2. Packets that were received from the peer and processed but not yet acked,
> 3. the total number of paths considered active, which some implementation use to decide when to increase the Max Path ID.

Packet handling as mentioned in point 1 and 2 seems actually partly orthogonal here. If you have send the path abandon frame, you should probably not have any outstanding packets that need to be acked on that path. However, you should also never just forgot about outstanting/non-acked packets. If you decide to remove state about non-ack packets, you need to declare them lost and resent on another path. I guess if you really want to sent a path abandon frame even though you have outstanding packets, you should probably declare them lost and resent on another path right away.

About your point 2, I don't think there is an issue here and I'm not sure when this case could/would happen. If you have received packets, I guess you should first send the ACK and then the path abandon. Of course you can also send ACKs on other paths still.

The important part is about packet handling of new incoming packets and you should be able to process those for at least 3 PTOs after sending the path abandon frame. This means, you need to remember the path ID and corresponding CID(s) as well as the packet number space state. However, as soon as you remove the CID(s), you cannot process any new incoming packets anymore and as such you can then also remove all packet number space state as well for received packets. For out-going packets, it's really on you if you remove state even though there are ACKs missing; however, as I said above that means you have to retransmit on another paths.

The interesting case is actually your point 3. If you remove path state that means you remove the path ID and will consider it as not active anymore. As such you could send a new max path ID. However, as the other end is not sending you a path abandon frame it will consider the path ID as active and not increase its max path ID. Respectively, you might not be able to open a new path. But I guess that's what we have the path blocked frame for. So I guess there is no problem either way.

So should we maybe say this to be more clear:

> However, an endpoint SHOULD process incoming packets for the abandoned Path ID and corresponding connection IDs for at least 3 PTOs after sending the PATh_ABANDON frame.

{{spurious-stateless-reset}} and to be able to acknowledge the
last packets received from the peer as specified in {{ack-after-abandon}}.

If a peer sends a PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it might not be able to remove path state.
It is left to the implementation to handle this unexpected

> behavior as it does not impact interoperability. If the endpoint is no longer 

Author: ```suggestion
behavior as it does not impact interoperability. If the endpoint is no longer
```

{{spurious-stateless-reset}} and to be able to acknowledge the
last packets received from the peer as specified in {{ack-after-abandon}}.

This means path state removal is supposed to be triggered by the reception of a
PATH_ABANDON frame. If a peer sends an PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it either might not be able to remove path state or
could use additional timers, e.g. after receiving the corresponding ACK frame,
or other conditions like a maximum number of retransmissions to remove state.

> If path state is remove without receiving a PATH_ABANDON frame this could lead

Author: ```suggestion
If path state is removed without receiving a PATH_ABANDON frame this could lead
```

> This means path state removal is supposed to be triggered by the reception of a
> PATH_ABANDON frame. If a peer sends a PATH_ABANDON frame but never receives
> a corresponding PATH_ABANDON frame, it either might not be able to remove path state or
> could use additional timers, e.g. after receiving the corresponding ACK frame,
> or other conditions like a maximum number of retransmissions to remove state.
> If path state is removed without receiving a PATH_ABANDON frame this could lead
> to packet loss. It is left to the implementation to handle this unexpected

Author: ```suggestion
If a peer sends a PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it might not be able to remove path state.
It is left to the implementation to handle this unexpected
```

behavior as it does not impact interoperability.

> 

Reviewer: Less is more. I would remove extra text and just keep the following:

If a peer sends a PATH_ABANDON frame but never receives
a corresponding PATH_ABANDON frame, it ~~either~~ might not be able to remove path state ~~or
could use additional timers, e.g. after receiving the corresponding ACK frame,
or other conditions like a maximum number of retransmissions to remove state~~.
~~If path state is removed without receiving a PATH_ABANDON frame this could lead
to packet loss.~~ It is left to the implementation to handle this unexpected
behavior as it does not impact interoperability.
